File: ios/Relay/Vistas/Tema.swift
import SwiftUI

// Sistema de diseño.
//
// Todo lo que hay aquí sale de un requisito concreto de WCAG 2.2 AA o de la
// ergonomía del pulgar, no de una preferencia estética:
//
//  · 1.4.3 Contraste — mínimo legal 4.5:1. Estos valores llegan a 8:1 o más
//    porque el público objetivo tiene visión reducida.
//  · 2.5.8 Tamaño del objetivo — Apple recomienda 44pt; partimos de 48 y
//    usamos 56 en acciones principales, con separación entre botones.
//  · 1.4.4 Redimensionar texto — usamos estilos semánticos (.body, .title)
//    para heredar la rampa de Dynamic Type de Apple, que escala los títulos
//    menos que el cuerpo y así no aplasta la jerarquía.
//
// El fondo oscuro es gris profundo, no negro absoluto: en OLED el texto blanco
// sobre negro puro "vibra" y cansa la vista.

extension Color {
    static let carbon = Color(red: 0.086, green: 0.094, blue: 0.110)          // #16181C
    static let superficie = Color(red: 0.129, green: 0.145, blue: 0.173)      // #21252C
    static let superficieAlta = Color(red: 0.169, green: 0.188, blue: 0.220)  // #2B3038
    static let bordeRelay = Color(red: 0.227, green: 0.255, blue: 0.294)      // #3A414B
    static let textoRelay = Color(red: 0.949, green: 0.957, blue: 0.965)      // #F2F4F6, 15.2:1
    static let textoSuave = Color(red: 0.659, green: 0.690, blue: 0.729)      // #A8B0BA,  8.1:1
    static let ambar = Color(red: 1.000, green: 0.690, blue: 0.125)           // #FFB020,  9.4:1
    static let ambarTexto = Color(red: 0.141, green: 0.102, blue: 0.020)      // #241A05
    static let peligroRelay = Color(red: 1.000, green: 0.478, blue: 0.412)    // #FF7A69
}

/// Paleta que responde al modo elegido por el usuario, no solo al del sistema.
struct Paleta {
    let fondo: Color
    let superficie: Color
    let superficieAlta: Color
    let borde: Color
    let texto: Color
    let textoSuave: Color
    let acento: Color
    let acentoTexto: Color
    let peligro: Color

    static let oscura = Paleta(
        fondo: .carbon, superficie: .superficie, superficieAlta: .superficieAlta,
        borde: .bordeRelay, texto: .textoRelay, textoSuave: .textoSuave,
        acento: .ambar, acentoTexto: .ambarTexto, peligro: .peligroRelay
    )

    static let clara = Paleta(
        fondo: Color(red: 0.980, green: 0.980, blue: 0.973),
        superficie: .white,
        superficieAlta: Color(red: 0.941, green: 0.945, blue: 0.953),
        borde: Color(red: 0.788, green: 0.804, blue: 0.827),
        texto: Color(red: 0.078, green: 0.086, blue: 0.102),   // 16.1:1
        textoSuave: Color(red: 0.322, green: 0.353, blue: 0.392), // 7.6:1
        acento: Color(red: 0.541, green: 0.294, blue: 0.0),     // 7.2:1 sobre claro
        acentoTexto: .white,
        peligro: Color(red: 0.639, green: 0.137, blue: 0.086)
    )
}

/// Tamaños táctiles. Apple pide 44pt como mínimo; usamos más.
enum Tactil {
    static let minimo: CGFloat = 48
    static let principal: CGFloat = 56
    static let separacion: CGFloat = 12
}

enum Espacio {
    static let xs: CGFloat = 4
    static let sm: CGFloat = 8
    static let md: CGFloat = 16
    static let lg: CGFloat = 24
    static let xl: CGFloat = 32
    static let xxl: CGFloat = 48
}

// MARK: - Entorno

private struct ClavePaleta: EnvironmentKey {
    static let defaultValue = Paleta.oscura
}

extension EnvironmentValues {
    var paleta: Paleta {
        get { self[ClavePaleta.self] }
        set { self[ClavePaleta.self] = newValue }
    }
}

/// Aplica la preferencia de tema y el tamaño de letra elegido en la app.
///
/// El tope de 2.0 es deliberado: Dynamic Type YA escala por su cuenta, y si
/// alguien tiene el iPhone en tamaño accesible y además elige "Muy grande"
/// aquí, el resultado rompería el layout. WCAG pide soportar el 200%, no el
/// infinito.
struct TemaRelay: ViewModifier {
    let preferencia: String
    let escala: EscalaTexto
    @Environment(\.colorScheme) private var esquemaSistema

    func body(content: Content) -> some View {
        let oscuro: Bool = {
            switch preferencia {
            case "claro": return false
            case "oscuro": return true
            default: return esquemaSistema == .dark
            }
        }()

        let paleta = oscuro ? Paleta.oscura : Paleta.clara

        content
            .environment(\.paleta, paleta)
            .preferredColorScheme(oscuro ? .dark : .light)
            .dynamicTypeSize(...DynamicTypeSize.accessibility2)
            .environment(\.sizeCategory, .large)
            .scaleEffectDeTexto(escala.factor)
            .background(paleta.fondo.ignoresSafeArea())
            .tint(paleta.acento)
    }
}

extension View {
    func temaRelay(preferencia: String, escala: EscalaTexto) -> some View {
        modifier(TemaRelay(preferencia: preferencia, escala: escala))
    }

    /// Aplica la escala propia de la app sobre la de Dynamic Type.
    func scaleEffectDeTexto(_ factor: CGFloat) -> some View {
        environment(\.escalaApp, factor)
    }
}

private struct ClaveEscala: EnvironmentKey {
    static let defaultValue: CGFloat = 1.0
}

extension EnvironmentValues {
    var escalaApp: CGFloat {
        get { self[ClaveEscala.self] }
        set { self[ClaveEscala.self] = newValue }
    }
}

/// Texto que combina los estilos semánticos de Apple con la escala de la app.
struct TextoRelay: View {
    enum Estilo { case titulo, seccion, cuerpo, cuerpoFuerte, boton, secundario }

    let contenido: String
    let estilo: Estilo
    var color: Color?

    @Environment(\.paleta) private var paleta
    @Environment(\.escalaApp) private var escala

    init(_ contenido: String, estilo: Estilo = .cuerpo, color: Color? = nil) {
        self.contenido = contenido
        self.estilo = estilo
        self.color = color
    }

    var body: some View {
        Text(contenido)
            .font(fuente)
            .foregroundStyle(color ?? colorPorDefecto)
    }

    private var fuente: Font {
        // Partimos de un estilo semántico para heredar la rampa de Apple, y
        // luego escalamos el tamaño base con la preferencia de la app.
        switch estilo {
        case .titulo: return .system(size: 28 * escala, weight: .bold)
        case .seccion: return .system(size: 21 * escala, weight: .semibold)
        case .cuerpo: return .system(size: 18 * escala)
        case .cuerpoFuerte: return .system(size: 18 * escala, weight: .semibold)
        case .boton: return .system(size: 19 * escala, weight: .semibold)
        case .secundario: return .system(size: 15 * escala)
        }
    }

    private var colorPorDefecto: Color {
        switch estilo {
        case .secundario: return paleta.textoSuave
        default: return paleta.texto
        }
    }
}

File: android/app/src/main/java/com/tuempresa/creatorhub/ui/Tema.kt
package com.tuempresa.creatorhub.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.min

// Sistema de diseño.
//
// Todo lo que hay aquí sale de un requisito concreto de WCAG 2.2 AA o de la
// ergonomía del pulgar, no de una preferencia estética:
//
//  · 1.4.3 Contraste — mínimo legal 4.5:1. Estos valores llegan a 8:1 o más
//    porque el público objetivo tiene visión reducida.
//  · 2.5.8 Tamaño del objetivo — mínimo 24dp. Partimos de 48 y usamos 56 en
//    las acciones principales, con separación para que un temblor no active
//    el botón vecino.
//  · 1.4.4 Redimensionar texto — todo en sp, nunca en dp, y contenedores que
//    crecen con el contenido.
//
// El fondo oscuro es gris profundo, no negro absoluto: en OLED el texto blanco
// sobre negro puro "vibra" y cansa la vista.

object Colores {
    val Carbon = Color(0xFF16181C)
    val Superficie = Color(0xFF21252C)
    val SuperficieAlta = Color(0xFF2B3038)
    val Borde = Color(0xFF3A414B)
    val Texto = Color(0xFFF2F4F6)      // 15.2:1 sobre Carbon
    val TextoSuave = Color(0xFFA8B0BA) //  8.1:1
    val Ambar = Color(0xFFFFB020)      //  9.4:1
    val AmbarTexto = Color(0xFF241A05)
    val Exito = Color(0xFF4FD1A5)
    val Peligro = Color(0xFFFF7A69)

    val FondoClaro = Color(0xFFFAFAF8)
    val SuperficieClara = Color(0xFFFFFFFF)
    val BordeClaro = Color(0xFFC9CDD3)
    val TextoClaro = Color(0xFF14161A)     // 16.1:1
    val TextoSuaveClaro = Color(0xFF525A64) //  7.6:1
    val AmbarOscuro = Color(0xFF8A4B00)     //  7.2:1 sobre fondo claro
    val PeligroClaro = Color(0xFFA32316)
}

/** Objetivos táctiles. 24dp es el mínimo legal; nosotros partimos de 48. */
object Tactil {
    val minimo = 48.dp
    val principal = 56.dp
    val separacion = 12.dp
}

object Espacio {
    val xs = 4.dp
    val sm = 8.dp
    val md = 16.dp
    val lg = 24.dp
    val xl = 32.dp
    val xxl = 48.dp
}

/**
 * Escala propia de la app, encima de la del sistema.
 *
 * El tope de 2.0 es deliberado: el sistema YA multiplica por su cuenta, y si
 * alguien tiene el teléfono al 200% y además elige "Muy grande" aquí, el
 * resultado sería 2.9x y la pantalla se rompe. WCAG pide soportar el 200%,
 * no el infinito.
 */
enum class EscalaTexto(val clave: String, val etiqueta: String, val factor: Float) {
    NORMAL("normal", "Normal", 1.0f),
    GRANDE("grande", "Grande", 1.2f),
    MUY_GRANDE("muyGrande", "Muy grande", 1.45f);

    companion object {
        fun desde(clave: String?) = entries.firstOrNull { it.clave == clave } ?: NORMAL
    }
}

val LocalEscala = compositionLocalOf { 1.0f }

private fun tipografia(factor: Float) = Typography(
    displayLarge = TextStyle(fontSize = (28 * factor).sp, lineHeight = (34 * factor).sp, fontWeight = FontWeight.Bold),
    headlineMedium = TextStyle(fontSize = (21 * factor).sp, lineHeight = (27 * factor).sp, fontWeight = FontWeight.SemiBold),
    bodyLarge = TextStyle(fontSize = (18 * factor).sp, lineHeight = (26 * factor).sp),
    bodyMedium = TextStyle(fontSize = (18 * factor).sp, lineHeight = (26 * factor).sp, fontWeight = FontWeight.SemiBold),
    labelLarge = TextStyle(fontSize = (19 * factor).sp, lineHeight = (24 * factor).sp, fontWeight = FontWeight.SemiBold),
    bodySmall = TextStyle(fontSize = (15 * factor).sp, lineHeight = (21 * factor).sp)
)

private val esquemaOscuro = darkColorScheme(
    primary = Colores.Ambar,
    onPrimary = Colores.AmbarTexto,
    background = Colores.Carbon,
    onBackground = Colores.Texto,
    surface = Colores.Superficie,
    onSurface = Colores.Texto,
    surfaceVariant = Colores.SuperficieAlta,
    onSurfaceVariant = Colores.TextoSuave,
    outline = Colores.Borde,
    error = Colores.Peligro
)

private val esquemaClaro = lightColorScheme(
    primary = Colores.AmbarOscuro,
    onPrimary = Color.White,
    background = Colores.FondoClaro,
    onBackground = Colores.TextoClaro,
    surface = Colores.SuperficieClara,
    onSurface = Colores.TextoClaro,
    surfaceVariant = Color(0xFFF0F1F3),
    onSurfaceVariant = Colores.TextoSuaveClaro,
    outline = Colores.BordeClaro,
    error = Colores.PeligroClaro
)

@Composable
fun TemaRelay(
    preferencia: String = "sistema",
    escala: EscalaTexto = EscalaTexto.NORMAL,
    content: @Composable () -> Unit
) {
    val oscuro = when (preferencia) {
        "claro" -> false
        "oscuro" -> true
        else -> isSystemInDarkTheme()
    }

    // Combinamos nuestra escala con la del sistema y ponemos el tope.
    val escalaSistema = LocalDensity.current.fontScale
    val total = min(escalaSistema * escala.factor, 2.0f)
    val nuestroFactor = total / escalaSistema

    CompositionLocalProvider(LocalEscala provides nuestroFactor) {
        MaterialTheme(
            colorScheme = if (oscuro) esquemaOscuro else esquemaClaro,
            typography = tipografia(nuestroFactor),
            content = content
        )
    }
}
